bpf/bpf_lxc.c

```diff
@@ -1963,6 +1963,82 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 	return ret;
 }
 
+static __always_inline bool
+ipv4_to_endpoint_is_hairpin_flow(struct __ctx_buff *ctx, struct iphdr *ip4)
+{
+	struct ipv4_ct_tuple tuple = {};
+	struct lb4_backend *backend;
+	struct ct_entry *entry;
+	struct ct_map *map;
+	int err, l4_off;
+	__be32 pod_ip, service_ip;
+	__be16 client_port, backend_port, service_port;
+
+	/* Extract the tuple from the packet so we can freely access addrs and ports.
+	 * All values are in network byte order.
+	 */
+	err = lb4_extract_tuple(ctx, ip4, ETH_HLEN, &l4_off, &tuple);
+	if (IS_ERR(err))
+		return false;
+
+	/* If the packet originates from a regular, non-loopback address, it will look
+	 * like service_ip:client_port -> pod_ip:service_port.
+	 *
+	 * In order to determine whether the packet has been hairpinned, we need to
+	 * obtain the backend (listen) port first, requiring a CT lookup with the
+	 * TUPLE_F_SERVICE flag, followed by a backend lookup. After this, the regular
+	 * CT TUPLE_F_OUT lookup can proceed.
+	 */
+	service_ip = tuple.saddr;
+	pod_ip = tuple.daddr;
+	client_port = tuple.sport;
+	service_port = tuple.dport;
+
+	tuple.daddr = service_ip;
+	tuple.saddr = pod_ip;
+	tuple.dport = client_port;
+	tuple.sport = service_port;
+
+	tuple.flags = TUPLE_F_SERVICE;
+
+	map = select_ct_map4(ctx, CT_SERVICE, &tuple);
```

Comment:
Pretty sure this should be `get_ct_map4()`. There's no chance that Cluster-IDs could play into the mix here - not in theory, and not in practice (as `cil_to_container()` clears the `skb->cb` and we thus know it doesn't contain a `cluster_id`).

(same for the second user below)

Reply:
Good catch, thank you! Now addressed.